File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        unordered_map<char,int>mp;
        int start = 0;
        int ans = 0;

        for(int e=0;e<n;e++){
            if(mp.find(s[e]) != mp.end()) start = max(start, mp[s[e]] + 1);
            mp[s[e]] = e;
            ans = max(ans,e-start+1);
        }

        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = {}
        start = 0
        maxlen = 0

        for i,ch in enumerate(s):
            if ch in seen:
                start = max(start,seen[ch]+1)
            seen[ch] = i
            maxlen = max(maxlen,i-start+1)
        return maxlen